### PR TITLE
generate random password for initial okta login with heroku

### LINKED ISSUE #13768: Okta password rules have changed. JHipster + Heroku needs updating.
##### **Overview of the issue**

We have the following question in our Heroku integration, but if I enter `Password1`, it’s accepted, and I can’t log in afterward. So it's not working.

```
? Initial password for the JHipster Admin user. Password requirements: at least 8 characters, 
a lowercase letter, an uppercase letter, a number, no parts of your username.
```

I'm trying to find out what our new rules are.

##### **Motivation for or Use Case**

Password rules should reflect a password that will work with Okta.

##### **Reproduce the error**

```
jhipster jdl blog-oauth2
jhipster heroku
```

Then use `Password1` for the password.

##### **JHipster Version(s)**

v7 beta 1

### PR DESCRIPTION
**Pull request by atomfrede**
This PR generates a random password for the initial okta login when using heroku. The password is printed to the console afterwards (or can be checked in configuration as before). 

closes #13768

---

Please make sure the below checklist is followed for Pull Requests.

- [ ] [All continuous integration tests](https://github.com/jhipster/generator-jhipster/actions) are green
- [ ] Tests are added where necessary
- [ ] [jhipster-online](https://github.com/jhipster/jhipster-online) was updated if necessary
- [ ] Documentation is added/updated where necessary
- [ ] Coding Rules & Commit Guidelines as per our [CONTRIBUTING.md document](https://github.com/jhipster/generator-jhipster/blob/main/CONTRIBUTING.md) are followed

When you are still working on the PR, consider converting it to Draft (bellow reviewers) and adding `skip-ci` label, you can still see CI build result at your branch.

### CODE REVIEW

File: generators/heroku/index.js [hunk 1]
```diff
@@ -756,6 +737,12 @@ module.exports = class extends BaseBlueprintGenerator {
               }
               if (curlAvailable && jqAvailable) {
                 this.log(chalk.green('Running ./provision-okta-addon.sh to create all required roles and users to use with jhipster.'));
+                this.log(
+                  chalk.green.bold(
+                    `Going to provision okta. Use ${this.oktaAdminLogin}/${this.oktaAdminPassword} to login. You will need to change your password afterwards.`
```
Comment (mraible) [inline generators/heroku/index.js:742]: Is "Going to provision Okta." necessary here? I'd change it to something like:

```
Okta configured successfully! Use ${this.oktaAdminLogin}/${this.oktaAdminPassword} to sign in.
```

File: generators/heroku/index.js [hunk 2]
```diff
@@ -830,7 +822,7 @@ module.exports = class extends BaseBlueprintGenerator {
                 } catch (err) {
                   this.log(
                     chalk.red(
-                      'Failed to execute ./provision-okta-addon.sh. Make sure to setup okta according to https://www.jhipster.tech/heroku/.'
+                      'Failed to execute ./provision-okta-addon.sh. Make sure to setup Okta according to https://www.jhipster.tech/heroku/.'
```
Comment (mraible) [inline generators/heroku/index.js:825]:
```suggestion
                      'Failed to execute ./provision-okta-addon.sh. Make sure to set up Okta according to https://www.jhipster.tech/heroku/.'
```

File: generators/heroku/index.js [hunk 3]
```diff
@@ -803,6 +790,11 @@ module.exports = class extends BaseBlueprintGenerator {
             this.log(chalk.yellow(`After application modification, redeploy it with\n\t${chalk.bold('jhipster heroku')}`));
 
             if (this.useOkta) {
+              this.log(
+                chalk.green.bold(
+                  `Going to provision Okta. Use ${this.oktaAdminLogin}/${this.oktaAdminPassword} to login. You will need to change your password afterwards.`
```
Comment (mraible) [inline generators/heroku/index.js:795]: Same as above.

**Review by mraible — CHANGES_REQUESTED** [review]

File: generators/heroku/index.js (hunk 1 above)
Comment (atomfrede) [reply generators/heroku/index.js:742]: In fact it is not yet configured successfully, but I agree, it sounds better,

Comment (mraible) [reply generators/heroku/index.js:742]: I wasn't aware of that. It's best not to give a false positive. Is it possible to emit the user/pass at the end of the process? 

One issue we noticed with the Okta CLI, is that we need to pause (or something) to make sure the user didn't see the password scroll by (and clear the console).

Comment (mraible) [reply generators/heroku/index.js:742]: Advice from a co-worker:

> I think it's fine, but you might want to prompt a "make note of this password and continue" if possible
> You could also pop a browser with a password reset link

Comment (mraible) [reply generators/heroku/index.js:742]: How do you generate the password reset link? 

You call our https://developer.okta.com/docs/reference/api/users/#reset-password API.

This returns an object like this:

```json
{
  "resetPasswordUrl": "https://${yourOktaDomain}/reset_password/XE6wE17zmphl3KqAPFxO"
}
```

Then you can replace `reset_password` with `welcome` (so you get our new password reset flow). 

Not sure if this is worth it, but figured I'd mention it.

Comment (atomfrede) [reply generators/heroku/index.js:742]: I think the current flow is fine. You login from jhipster and need to change your password directly. No need to do a dedicated reset step imho.

**Comment by atomfrede** [comment]
Added a `confirm` prompt to make sure the user is aware of the generated password. The hint about login is now displayed after the script has been executed.

**Comment by mraible** [comment]
@atomfrede I tried to test this with your branch, but it fails to deploy to Heroku with the "build on Heroku" option. I think this is because jhipster-dependencies' SNAPSHOT is unavailable, I'm not sure.

```
remote: -----> Executing Maven
remote:        $ ./mvnw -Pprod,heroku -DskipTests clean dependency:list install

remote:        [INFO] Scanning for projects...

remote:        [INFO] Downloading from ossrh-snapshots: https://oss.sonatype.org/content/repositories/snapshots/tech/jhipster/jhipster-dependencies/7.0.0-SNAPSHOT/maven-metadata.xml

remote:        [INFO] Downloading from ossrh-snapshots: https://oss.sonatype.org/content/repositories/snapshots/tech/jhipster/jhipster-dependencies/7.0.0-SNAPSHOT/jhipster-dependencies-7.0.0-SNAPSHOT.pom

remote:        [ERROR] [ERROR] Some problems were encountered while processing the POMs:
remote:        [ERROR] Non-resolvable import POM: Could not find artifact tech.jhipster:jhipster-dependencies:pom:7.0.0-SNAPSHOT in ossrh-snapshots (https://oss.sonatype.org/content/repositories/snapshots/) @ line 104, column 25
remote:        [ERROR] 'dependencies.dependency.version' for tech.jhipster:jhipster-framework:jar is missing. @ line 116, column 21

remote:        [ERROR] 'dependencies.dependency.version' for javax.annotation:javax.annotation-api:jar is missing. @ line 120, column 21
remote:        [ERROR] 'dependencies.dependency.version' for org.springframework.boot:spring-boot-starter-cache:jar is missing. @ line 124, column 21
```

I tried to run `jhipster heroku` again to build locally, but it fails too.

```
Configuring Heroku
No new Database addon created
✖ Error: Command failed: heroku buildpacks:add heroku/java --app shrouded-mountain-34941
 ›   Error: The buildpack heroku/java is already set on your app.

    at ChildProcess.exithandler (child_process.js:308:12)
    at ChildProcess.emit (events.js:315:20)
    at maybeClose (internal/child_process.js:1048:16)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:288:5) {
  killed: false,
  code: 1,
  signal: null,
  cmd: 'heroku buildpacks:add heroku/java --app shrouded-mountain-34941',
  stdout: '',
  stderr: ' ›   Error: The buildpack heroku/java is already set on your app.\n'
```

One thing I noticed is the Maven command that's used on Heroku:

```
./mvnw -Pprod,heroku -DskipTests clean dependency:list install
```

It seems like it could be a bit shorter, like:

```
./mvnw -Pprod,heroku -DskipTests package
```

**Comment by mraible** [comment]
Related: building locally seems to run all the tests too. It's probably a good practice, but not sure if it's necessary. 

```
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  03:35 min
[INFO] Finished at: 2021-02-02T08:36:14-07:00
[INFO] ------------------------------------------------------------------------
```

It'd be cool if you could build and deploy a JHipster app to Heroku in 2 minutes!

**Comment by mraible** [comment]
@atomfrede I got it to work. I forgot you needed to run `npm link generator-jhipster` in the project after generation.

Here's what it currently says:

```
? You are using OAuth 2.0. Do you want to use Okta as your identity provider it yourself? When you choose Okta, the auto
mated configuration of users and groups requires cURL and jq. Yes, provision the Okta add-on
? Login (valid email) for the JHipster Admin user: matt.raible@okta.com
? Take note of this password! You will need it on your first login: 28be10ea4c0501b162c0f6dfdb0cf64a49ee767b Yes
```

Can we change it to the following?

```
? You are using OAuth 2.0. Do you want to use Okta? When you choose Okta, the automated configuration of users and groups requires cURL and jq. Yes, provision the Okta add-on
? Enter a valid email for the JHipster Admin user: matt.raible@okta.com
? Take note of this password! You will need it to sign in: 28be10ea4c0501b162c0f6dfdb0cf64a49ee767b Yes
```

I could also see changing the initial question to `Do you want to use Okta as your identity provider?` The reason I removed it is I don't think developers are that familiar with the term 'identity provider', but I could be wrong.

It'd also be cool if the password could be a different color so it's more noticeable. You probably don't want it to be blue because then the user might think "Yes" is part of it. Maybe yellow?

Screen Shot 2021-02-02 at 8 42 55 AM

At the end, there's a NaN. Not sure what that should be.

Screen Shot 2021-02-02 at 8 48 34 AM

I'm using the `blog-oauth2` JDL to test. It looks like Hazlecast might be preventing a successful startup.

Screen Shot 2021-02-02 at 8 51 30 AM

**Comment by atomfrede** [comment]
> @atomfrede I got it to work. I forgot you needed to run `npm link generator-jhipster` in the project after generation.
> 
> Here's what it currently says:
> 
> ```
> ? You are using OAuth 2.0. Do you want to use Okta as your identity provider it yourself? When you choose Okta, the auto
> mated configuration of users and groups requires cURL and jq. Yes, provision the Okta add-on
> ? Login (valid email) for the JHipster Admin user: matt.raible@okta.com
> ? Take note of this password! You will need it on your first login: 28be10ea4c0501b162c0f6dfdb0cf64a49ee767b Yes
> ```
> 
> Can we change it to the following?
> 
> ```
> ? You are using OAuth 2.0. Do you want to use Okta? When you choose Okta, the automated configuration of users and groups requires cURL and jq. Yes, provision the Okta add-on
> ? Enter a valid email for the JHipster Admin user: matt.raible@okta.com
> ? Take note of this password! You will need it to sign in: 28be10ea4c0501b162c0f6dfdb0cf64a49ee767b Yes
> ```
> 
> I could also see changing the initial question to `Do you want to use Okta as your identity provider?` The reason I removed it is I don't think developers are that familiar with the term 'identity provider', but I could be wrong.
> 

Sure.

> It'd also be cool if the password could be a different color so it's more noticeable. You probably don't want it to be blue because then the user might think "Yes" is part of it. Maybe yellow?

Need to check if we can use different colors for the prompts

> 
> Screen Shot 2021-02-02 at 8 42 55 AM
> 
> At the end, there's a NaN. Not sure what that should be.
> 
> Screen Shot 2021-02-02 at 8 48 34 AM
> 
> I'm using the `blog-oauth2` JDL to test. It looks like Hazlecast might be preventing a successful startup.
> 
> Screen Shot 2021-02-02 at 8 51 30 AM

The `NaN` is wrong. Will check. It should be `EMAIL/PASSWORD`.

**Comment by mraible** [comment]
I tried again without Hazelcast and it worked. I like the "reset your password on first use flow!"

Besides the other edits I suggested, it might be cool to spit out the credentials at the end. I'm not sure we need to tell them they'll need to change their password afterward.

```
Okta configured successfully!

Use $email/$password to sign in. 
```

**Comment by atomfrede** [comment]
It now looks like this

image

Screenshot-20210202222434-887x151

File: generators/heroku/index.js [hunk 4]
```diff
@@ -226,7 +228,7 @@ module.exports = class extends BaseBlueprintGenerator {
             type: 'list',
             name: 'useOkta',
             message:
-              'You are using OAuth 2.0. Do you want to use Okta as your identity provider it yourself? When you choose Okta, the automated configuration of users and groups requires cURL and jq.',
+              'You are using OAuth 2.0. Do you want to use Okta? When you choose Okta, the automated configuration of users and groups requires cURL and jq. Yes, provision the Okta add-o',
```
Comment (mraible) [inline generators/heroku/index.js:231]: Did you mean to add the new part to the end of this sentence?

**Review by mraible — APPROVED** [review]
@atomfrede I would change the end messages to capitalize JHipster. There's a `jhipster` in there.

**Comment by atomfrede** [comment]
Bounty claimed https://opencollective.com/generator-jhipster/expenses/33184

**Comment by pascalgrimaud** [comment]
@atomfrede : approved

**Pull request merged**